include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -672,27 +672,35 @@ template <typename _Event, typename... _Args>
 class __future : private std::tuple<_Args...>
 {
     _Event __my_event;
+    bool __waited_for_result = false;
 
     template <typename _T>
     constexpr auto
     __wait_and_get_value(const sycl::buffer<_T>& __buf)
     {
         //according to a contract, returned value is one-element sycl::buffer
-        return __buf.get_host_access(sycl::read_only)[0];
+        auto __val = __buf.get_host_access(sycl::read_only)[0];
+        __set_waited_for_result();
+        return __val;
     }
 
     template <typename _ExecutionPolicy, typename _T>
     constexpr auto
     __wait_and_get_value(const __result_and_scratch_storage<_ExecutionPolicy, _T>& __storage)
     {
-        return __storage.__wait_and_get_value(__my_event);
+        auto __val = __storage.__wait_and_get_value(__my_event);
```

Comment:
Similar comment here about unneeded waits in case we've waited before.

Reply:
https://github.com/oneapi-src/oneDPL/pull/1747#discussion_r1715079486

Reply:
as result, all `waited_for_result` stuff has been removed to reduce the amount of changes.

Reply:
@julianmi may we resolve this conversation?